### ISSUE TITLE
On macOS, pciutils only shows devices attached to the PCH

### ISSUE THREAD

**Issue by jyavenard** [pciutils/pciutils#4]
I have a X299 board (running macOS 10.13)
pciutils only shows the devices connected to Intel's X299 PCH.

None of the PCI devices running on the CPU's PCIe lanes are showing.

**Comment by gollux** [comment]
Unfortunately, the MacOS back-end for the pciutils is rather primitive. In particular, it is unable to ask the kernel for the list of available buses/devices, so lspci has to resort to a trivial bus scanning strategy. You can try 'lspci -M', it might help.

The MacOS back-end is currently unmaintained. If anyone is willing to take care of it, I will be happy.

**Comment by joevt** [comment]
Fixed in my fork. See comment at https://github.com/pciutils/pciutils/issues/111#issuecomment-1407659410

**Comment by pali** [comment]
lib/darwin.c backend uses pci_generic_scan() function for scanning of devices and this function scans _only_ domain 0. So devices from other domains are not detected by lib/darwin.c


@gollux What about renaming `pci_generic_scan()` function to e.g. `pci_generic_scan_domain0()` to make it explicit what this function is scanning? It could prevent confusion like in this issue.

**Comment by joevt** [comment]
@pali , Yes, `pci_generic_scan()` scans **only** domain 0 but it doesn't scan **all** of domain 0. It scans only buses attached to bus 00 so `pci_generic_scan_domain0()` is not specific enough. Besides, I don't see how changing a function name is going to prevent a user from being confused.

My fork changes the `darwin` access method to scan all buses of the domain and adds a `darwin3` access method for doing the same for a specific domain. https://github.com/pciutils/pciutils/issues/111#issuecomment-1407659410

**Comment by pali** [comment]
> @pali , Yes, `pci_generic_scan()` scans **only** domain 0 but it doesn't scan **all** of domain 0. It scans only buses attached to bus 00 so `pci_generic_scan_domain0()` is not specific enough.

Yea, I totally forgot about multiroot PCIe controllers with each PCIe Root Port on separate bus (but on same domain as they on the one same PCIe controller). For these controllers even they have all devices on one domain, this function does not discover all devices. So better name would be `pci_generic_scan_domain0_bus0_subtree()` or something like that.

> Besides, I don't see how changing a function name is going to prevent a user from being confused.

Well, I was already confused that I though that this function named `pci_generic_scan()` would scan and discover _all_ devices.

**Comment by gollux** [comment]
Re naming of functions: I do not see that the name claims that the function scans every bus in the visible universe. If you find the behavior of the functions confusing, just add a comment. No need to call it silly names.
